### PR TITLE
add documentation about the federation and node policies

### PR DESCRIPTION
**Pull request by mishaschwartz**
This is a copy of the first draft of this document (originally a google doc).
Now that this repo exists, this is a good place to put this sort of policy documentation.

Please comment and continue the discussion about changes to this documentation here instead of the google doc.

### CODE REVIEW

**Review by mishaschwartz — COMMENTED** [review]
Here are all the remaining comments from the google doc that are still in discussion. I'm just porting them over here:

File: docs/federation.md [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+- Deployment of software stack: Each Node should deploy the DACCS software stack. Node Administrators are required to
+  upgrade the software stack to the latest DACCS stack within two weeks of the updates release date.
+- Node maintenance: Node Administrators accept that joining the Network implies that they are responsible for
+  provisioning, funding and maintaining their own Node.
+- Node security: Node Administrators are responsible for maintaining secure access to their node and for coordinating
+  with
+  the hosting institution regarding any security related considerations prior to the deployment of the Node. Admin
+  access
+  to the Node must be restricted to listed Node Administrators.
+- Downtime policy: A Node should strive to maintain downtime of no more than 5%. If a Node needs to be taken offline for
```
[the remark this answers is not in the document]
Comment (tlvu) [reply docs/federation.md:27]: If we notify the users before hand, can we have a downtime of more than 5%?

Example when changing our production machine, we needed a downtime of 1 day. 

We sometime also have unscheduled power outage as well that lasts way more than just 1 hour.

Currently if an event is scheduled, we notify all our Ouranos Jupyter users.  In a federation, should we notify all users in the federation?

File: docs/federation.md [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+
+- Privacy: No Personal Identifying Information belonging to a User can be made accessible to anyone except the User in
+  question, the Node Manager and the DACCS Executive Committee (“Executive Committee”) without explicit written consent
+  of
+  the User in question.
+- No User Generated Data can be made accessible to another User without the User’s permission. without the user’s
+  permission. User Generated Data can not be made accessible outside the Network without the User’s permission.
+- Back-up of User Generated Data: The Node Administrator should have a mechanism in place to regularly back up a user’s
+  main
+  working directory, and have a mechanism in place to retrieve the back-up data upon a user’s request.
```
[the remark this answers is not in the document]
Comment (tlvu) [reply docs/federation.md:61]: Ouranos currently do not provide any data backup to our Jupyter users.  Data retention is basically as long as they are active Jupyter users.  Otherwise we can also delete their account to reclaim disk space.

Users are responsible to export their code or data output if they need to preserve them.

We also ask our users to be sensible to the amount of disk space, CPU, memory they use.  The platform is not an HPC.

Can each node have their own backup and data retention policy or all should follow the same policy?

**Comment by tlvu** [comment]
@huard @tlogan2000 You might want to review this document as well.

File: docs/federation.md [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+  provisioning, funding and maintaining their own Node.
+- Node security: Node Administrators are responsible for maintaining secure access to their node and for coordinating
+  with
+  the hosting institution regarding any security related considerations prior to the deployment of the Node. Admin
+  access
+  to the Node must be restricted to listed Node Administrators.
+- Downtime policy: A Node should strive to maintain downtime of no more than 5%. If a Node needs to be taken offline for
+  scheduled maintenance purposes, the Node Administrator must provide users at least 2-weeks notice.
+- New user signups: Node Administrators are responsible for the signups of users on their own Nodes. All new signup
+  requests should be verified and approved or rejected within 2 business days.
```
Comment (tlvu) [inline docs/federation.md:30]: This 2 business days can be a bit short, especially in a high season vacation.

File: docs/federation.md [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+- No User Generated Data can be made accessible to another User without the User’s permission. without the user’s
+  permission. User Generated Data can not be made accessible outside the Network without the User’s permission.
+- Back-up of User Generated Data: The Node Administrator should have a mechanism in place to regularly back up a user’s
+  main
+  working directory, and have a mechanism in place to retrieve the back-up data upon a user’s request.
+
+### Published Data
+
+Published Data must be discoverable and readable by all Users on a Node. Node Managers are responsible for ensuring that
+all published data is recoverable in case of data loss. Node managers should provide Users at least one-week notice
```
[the remark this answers is not in the document]
Comment (tlvu) [reply docs/federation.md:66]: Published data I assume we mean Thredds and Geoserver data?  If those are down, most likely it's the network, or electricity or hardware failure.   Will be hard to specify a recovery time-frame.  

Maybe we can say that if we foresee the downtime will be longer than X hours, we will notify the users, instead of leaving them in the dark.

**Review by tlvu — COMMENTED** [review]
A few other quick thoughts.

File: docs/federation.md (hunk 1 above)
[the remark this answers is not in the document]
Comment (mishaschwartz) [reply docs/federation.md:27]: Maybe we should clarify that this is unscheduled downtime. Scheduled downtime can of course be as long as needed as long as the users are warned beforehand (see next sentence). 

We should reword this to talk about having a policy to recover from unscheduled outages within X number of hours/days instead of in terms of a percentage

File: docs/federation.md (hunk 2 above)
[the remark this answers is not in the document]
Comment (mishaschwartz) [reply docs/federation.md:61]: Each node can have its own policy. We're trying to outline minimum requirements here.

> Data retention is basically as long as they are active Jupyter users.

Ok, but what happens if you're storing active jupyter workspaces on a disk and that disk fails? Do you have any backups or RAIDing of those disks?

File: docs/federation.md (hunk 4 above)
[the remark this answers is not in the document]
Comment (mishaschwartz) [reply docs/federation.md:66]: Published data is anything that is not editable by a user and is therefore relatively stable. This includes most data served by thredds/geoserver but doesn't include things like wps ouputs, data in user workspaces, etc.

> Maybe we can say that if we foresee the downtime will be longer than X hours, we will notify the users, instead of leaving them in the dark.

I like that idea. I'm happy to prioritize communication with the users over specific recovery policies that may be burdensome to achieve.

File: docs/federation.md (hunk 3 above)
Comment (mishaschwartz) [reply docs/federation.md:30]: What would you recommend?

Comment (tlvu) [reply docs/federation.md:30]: Maybe add the word "ideally" in front of the "2 business days"?

I like us to convey the idea that we will try hard to meet the level of service but it's not a warranty.  I fear this document will induce users to think this is our TOS (terms of service) as in a commercial contract.  I don't think we are at that level yet, especially we are all non profit organisations, providing the service for free to our users.

Same spirit for all other limits discussed in this document.

File: docs/federation.md (hunk 2 above)
[the remark this answers is not in the document]
Comment (tlvu) [reply docs/federation.md:61]: No backup but our disks are RAID.

**Comment by tlvu** [comment]
@fmigneault @matprov @eyvorchuk You might want to review this PR.

File: docs/federation.md (hunk 3 above)
Comment (mishaschwartz) [reply docs/federation.md:30]: There are two goals with this document: 

1. providing Node Administrators with a set of guidelines that we _recommend_ they follow. If they choose not to follow these, then _their node_ may not run as well as we'd like but that's ok.
2. providing Node Administrators with a set of guidelines that they _must_ follow. If they choose not to follow these, then the node will degrade the quality of the network itself. If they do not follow these then we may choose to remove their node from the network.

This is making me think that we should reorganize this document to clarify these two goals.

File: docs/federation.md (hunk 2 above)
[the remark this answers is not in the document]
Comment (mishaschwartz) [reply docs/federation.md:61]: I'm happy to say:

- minimal requirement: RAID (or some other protection from disk failure)
- recommended requirement: full backups with a defined recovery policy

what do you think?

[the remark this answers is not in the document]
Comment (tlvu) [reply docs/federation.md:61]: Good for me.

File: docs/federation.md (hunk 3 above)
Comment (tlvu) [reply docs/federation.md:30]: Agreed. In the current state, I had the wrong impression all the guidelines are "must follow".

**Comment by eyvorchuk** [comment]
I don't think I have anything to add that hasn't already been mentioned, but I'm also going to ask @jameshiebert to review this.

File: docs/federation.md (hunk 4 above)
[the remark this answers is not in the document]
Comment (dchandan) [reply docs/federation.md:66]: Why are we saying that published data must be discoverable and readable "by all users on a node" and not "all users on the network?

[the remark this answers is not in the document]
Comment (mishaschwartz) [reply docs/federation.md:66]: We didn't want to step on a node manager's decision to keep some data private for their own users. Maybe we need to differentiate between:

published data: not editable by users (available to a subset of users in the network) 
public data: available to everyone (no authentication)

File: docs/federation.md [hunk 5]
```diff
@@ -0,0 +1,116 @@
...
+
+Node Administrators are expected to provide technical support to users of their Nodes in a timely manner.
+
+### User Data and Privacy
+
+- Privacy: No Personal Identifying Information belonging to a User can be made accessible to anyone except the User in
+  question, the Node Manager and the DACCS Executive Committee (“Executive Committee”) without explicit written consent
+  of
+  the User in question.
+- No User Generated Data can be made accessible to another User without the User’s permission. without the user’s
```
Comment (fmigneault) [inline docs/federation.md:57]: A few repeated words here.

File: docs/federation.md [hunk 6]
```diff
@@ -0,0 +1,116 @@
...
+
+## Defederation
+
+There are two ways a Node can be defederated from the Network: voluntary or forced.
+
+Voluntary removal is instigated by a Node Administrator who submits a request to the Executive Committee for the removal
+of the Node (that the Administrator manages) from the Network. In this case, the Executive Committee will work with the
+institution hosting the node and the local Node Administrator to work out a solution that would provide minimal
+disruption to the users of the Network and assist with the transfer of any unique data to another Node in the Network if
+desired.-. We encourage Node Administrators to provide as much notice as possible of the pending removal of the node to
```
Comment (fmigneault) [inline docs/federation.md:91]: typo `.-.`

File: docs/federation.md [hunk 7]
```diff
@@ -0,0 +1,116 @@
...
+- New user signups: Node Administrators are responsible for the signups of users on their own Nodes. All new signup
+  requests should be verified and approved or rejected within 2 business days.
+- Node configuration: The DACCS deployment stack consists of core components and optional components. A Node can enable
+  or
+  disable non-core components as per the decisions of the Node Administrator. However, when a feature/component is to be
+  removed, adequate notice for the change should be given to the Node’s users.
+
+### Support and Documentation
+
+A Node should maintain a public web-page with:
```
Comment (mishaschwartz) [inline docs/federation.md:38]: I think we should remove this requirement. Node information will now be accessible through the website and the content on the site is dictated by what is in the node registry.

File: docs/federation.md [hunk 8]
```diff
@@ -2,83 +2,74 @@
 
 ## Introduction
 
-The DACCS Network (“Network”) is a federated network that consists of independently operated and maintained computing
-infrastructure (“Node”) that deploy the DACCS software stack and connect with one another under the framework laid out
+The Marble Network ("Network") is a federated network that consists of independently operated and maintained computing
+infrastructure ("Node") that deploy the Marble software stack and connect with one another under the framework laid out
```
Comment (fmigneault) [inline docs/federation.md:6]: I think some form of "DACCS" and "birdhouse" references should be kept here. 

Given that only the network uses "Marble", and that everywhere else (code, docs, etc.) refers to DACCS/birdhouse stack (including even this repo's name), the mention "Marble software stack" does not really provide any reference to users and adds more confusion to this already complex infrastructure.

Comment (mishaschwartz) [reply docs/federation.md:6]: I disagree actually. Other code/docs refer to DACCS or birdhouse inconsistently and the point of coming up with a name like "Marble" is to create a consistent name that we can use to refer to the product that we are creating.

Here is how we should think about this:

- Marble: the name of the product
- DACCS: the name of the organization developing the product
- birdhouse: the name of a piece of software used to deploy one part of the product (a single node)
- magpie, cowbird, finch, etc.: components pieces of the product

Comment (fmigneault) [reply docs/federation.md:6]: > birdhouse: the name of a piece of software used to deploy one part of the product 

Exactly as you said, hence why "the Marble software stack" doesn't make sense IMO. The software is birdhouse, not Marble. 
I agree with using Marble when referring to "The Marble Network".

**Review by huard — APPROVED** [review]

File: docs/federation.md (hunk 8 above)
Comment (mishaschwartz) [reply docs/federation.md:6]: > Exactly as you said, hence why "the Marble software stack" doesn't make sense IMO. The software is birdhouse, not Marble.

Birdhouse is used for many things (as we've discussed at length). Marble is only one use-case for the birdhouse-deploy scripts. We've discussed in the past that birdhouse-deploy is used for many projects, not just Marble.

If you would like another way to think about it: "the Marble software stack can be deployed using a specific configuration of the birdhouse-deploy software".

Comment (fmigneault) [reply docs/federation.md:6]: I don't see why birdhouse-deploy being used by many projects makes its name invalid when referring to what actually gets deployed. Birdhouse is the stack being deployed, and each node as its own configuration that must respect a preset defined by the Marble Network.

Otherwise, it's like saying that a Python-based program "X" runs the "X programming language" instead of "running X with Python". Python being used by many other programs is irrelevant in how it is applied in the case of "X".

Comment (mishaschwartz) [reply docs/federation.md:6]: Yes... 

We are "running Marble with birdhouse-deploy" in the same way that you can "run X with Python". Maybe in the future we'll run Marble with something else, maybe not. But the point is that Marble is a software product that is distinct from birdhouse. We do not need to specify the underlying mechanism for deployment here.

> What would be the most explicit is simply to have [birdhouse](https://github.com/bird-house). That way, readers don't have to guess which software stack we refer to, there is an actual link.

Yes, but not here. We're talking about the product in the abstract here. Documentation for deploying and hosting a node belongs elsewhere.

Comment (fmigneault) [reply docs/federation.md:6]: This documentation can be adjusted in the future as necessary if the stack evolves. For the moment, and I expect it to be for a long time, birdhouse-deploy is what is running. 

The whole point of this kind of documentation is to guide new node maintainers that would be interested in joining the network, not us that already know the stack. The first question anyone new will have is where this "Marble stack" is, so they can try it out. And then we would point them at birdhouse-deploy... instead of directly giving them the reference here right away. 

It's not like we are trying to form a secret organization. Why all this abstract mystery? Why not openly provide the link?

Comment (dchandan) [reply docs/federation.md:6]: I feel that the phrase "Marble software stack" makes sense because the stack can be based on something other than birdhouse in the future as pointed out by Misha. At present, the loose naming conventions from the past and the way the development was coordinated, have created an unfortunate situation that birdhouse, PAVICS, and now, DACCS/Marble have all come to mean vaguely similar things even though they are not. But we now have the opportunity to clarify the distinction between Marble and birdhouse and we should use that opportunity. 

Perhaps, we can add somewhere in the documentation that the "Marble software stack is currently using the community birdhouse deploy mechanism blah blah blah"?

Comment (dchandan) [reply docs/federation.md:6]: > The first question anyone new will have is where this "Marble stack" is, so they can try it out. And then we would point them at birdhouse-deploy... instead of directly giving them the reference here right away.

I think the website, that is due to go live soon, is going to have clear instructions on this. We can even add a link to that web page in this document.

**Comment by mishaschwartz** [comment]
Pulling this in for now. We can discuss further edits at a later date.
